reframe/frontend/executors/__init__.py

```diff
@@ -197,6 +205,21 @@ def _environ_supported(self, check, environ):
         else:
             return ret and check.supports_environ(environ.name)
 
+    def _retry_failed(self, checks, system):
+        while (self._stats.num_failures() and self._current_run <
+                                              self._max_retries):
+            self._current_run += 1
+            self._stats.next_run(self._current_run)
```

Comment:
I disagree passing the current run to the `next_run`. `next_run` should simply advance the run count and take no argument. If sb wants to check if everything is ok, should do by himself, i.e., statistics should offer read-only access to their internal run count and any caller could validate the status.

What you are doing here is basically a `set_run`, but I really don't see the purpose for that.

Reply:
You should also advance the statistics run counter after you get the failures so as not to have to pass `run=-2` later.

Reply:
OK, I will remove the `current_run` argument from `next_run`. However, if you think about giving read-only access to the internal run count of the `statistics`, then I would say that we should rather give read-only access to the internal run count of the `Runner`. In fact, the runner has the original and always correct count; so it would make much more sense that this is made accessible to others. Yet, you did not like that in my previous versions; so I guess, we should just remove any counter verification...

Reply:
I am not against the counter verification, and it's good to have such guards generally (btw, it should raise `AssertionError` since it denotes a bug). What I am against to is the interface. Looking at the interface, it's like having a counter and an increment operation on it, but instead of letting the counter increment itself, you pass it's new value (the one *you * count) and the counter then checks this against its own counter. Doesn't this sound weird? It would be much more natural to let the counter count the way it knows and you also count your own way and compare your values afterwards for sanity.

Reply:
If we adopt what you propose for a `set_run()`, we are violating the notion of setting a property. Theoretically, with a `set_run` one would expect to be able to set any run number, but here it would be just an increment. It is exactly as the example with the counter I gave you before; it's like calling `counter.inc(new_value)`, where the only allowed value for `new_value` would be the current plus one. It doesn't make any sense, because in this case the caller does the counting, and the counter does the sanity checking! It should be the other way around. So I definitely prefer the first one.

Reply:
OK, done.